Patterns: Add pattern name to document toolbar when editing in spotlight

scruffian opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
When a contentOnly pattern is being editing in the canvas, we should show the pattern name and a back button in the document toolbar.

<!-- In a few words, what is the PR actually doing? -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
This tells the user that they are editing a pattern, and gives them a mechanism to finish editing.

Alternative to #73088

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Updates the document toolbar to detect when a pattern is being edited.

## Testing Instructions
0. Enable the content only experiment
1. Add a contentOnly pattern to a page
2. Edit a pattern in the inspector by clicking "Edit section"
3. Notice that the document toolbar is updated and has a Back button
4. Check that the Back button finished the pattern edit.
5. Try the same with template parts and synched patterns


## Screenshots or screencast <!-- if applicable -->


https://github.com/user-attachments/assets/0bf7a6a7-d581-4f5d-946d-8f78d489e3c0

scruffian committed:
add back button

scruffian committed:
also use the entity name

ramonjd commented on packages/editor/src/components/document-bar/index.js:
```diff
@@ -133,11 +171,28 @@ export default function DocumentBar( props ) {
+
+	// Use pattern info if a pattern block is unlocked, otherwise use document/entity info
+	const title =
+		unlockedPatternInfo?.patternTitle ||
```
This is a great change, I like! 

Should it work for synced patterns and `"templateLock":"contentOnly"` blocks as well? Both receive content only treatment.

ramonjd replied:
> Should it work for synced patterns and "templateLock":"contentOnly" blocks as well? 

Not sure about those, but I updated this PR to work with synced patterns and template parts. Hope you don't mind!

ramonjd committed:
Refactor DocumentBar: Introduce useEditedSectionDetails hook for improved section detail retrieval

- Removed inline logic for fetching details of the currently edited content-only section.
- Added a new custom hook, useEditedSectionDetails, to encapsulate the logic for retrieving pattern and template part details.
- Updated DocumentBar to utilize the new hook for cleaner and more maintainable code.

ramonjd reviewed (approved):
This is behind an experiment. It works as it should. 

blocks with `templateLock` are currently not catered for but that can be a follow up if it's at all necessary.

Thank you!
